### DISCUSSION TITLE
Api with tenant

### DISCUSSION THREAD

**Discussion by meliora-solution** [JonPSmith/AuthPermissions.AspNetCore#101] [Q&A]
The only Example I can create for blazor app is Example 2.  Can I use Example3 for web api ?  Because that is the only way I know how to implement Authp in blazor application.

**Comment by meliora-solution** [comment]
so far so good. I can create database for webApi project from example 3.

**Comment by JonPSmith** [comment]
Hi @meliora-solution,

Yes, using webApi is the best way to build a blazor app that uses AuthP's backend code. You might find the AuthP's following documentation useful:

- [JWT Token refresh explained](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/JWT-Token-refresh-explained)
- [JWT Token configuration](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/JWT-Token-configuration)
- [Using JWT Tokens](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Using-JWT-Tokens)

Also, if you want to use AuthP's Roles / [Permissions](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Using-Permissions) to control access then you might find [this section](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Using-JWT-Tokens#extra-sending-the-current-users-permissions-to-the-front-end) about you can get the user's permissions in your blazor side. This allows your frontend only show the features that the current user can access.

**Reply by meliora-solution** [reply]
Thank you John,

I tried several ways using Authp for example3.

1. Using New Blazor SSR (server Side Rendering) for example 3.  I can create database but still has problems.
2. Using JWT Token. This is okay and no problem.
3. I want to try to use new feature in NET 8 Identity Endpoints.  It is much easier than JWT Token because I can choose Cookie, session Cookie or Bearer Token. What is new in NET 8  `https://devblogs.microsoft.com/dotnet/whats-new-with-identity-in-dotnet-8/`

If I am able to use authp with that web api Identity Endpoint, that is very good because it is easier.

4. There is another way . I am not sure if that is a good idea ? I read about the BFF pattern (Backend for FrontEnd).  I can use Identity Endpoint as BFF (as a gateway) to connect web api jwt token.

User > BFF > JWT Token.  

User can connect using cookie to BFF.  BFF can connect to webApi with jwt Token.  So jwt token is used for company to company.

I am not sure if that Is that a good idea ?

**Comment by meliora-solution** [comment]
Yes I can use the new feature of NET 8 Identity Endpoints and use cookie authentication or bearer Token.  So to use Authp , you dont have to use Jwt token. There is one thing you must be aware that EmailConfirmed in AspNetusers must be set to true or else it wont work.

I have converted the mvc controller GetTenatnList with permission Example3Permission.TenantList to api controller.
```
  [HttpGet]
  [Route("tenantlist")]
  [HasPermission(Example3Permissions.TenantList)]
  public async Task<IActionResult> GetTenantList()
  {
      var tenantNames = await SingleLevelTenantDto.TurnIntoDisplayFormat(_authTenantAdmin.QueryTenants())
          .OrderBy(x => x.TenantName)
          .ToListAsync();

         return Ok(tenantNames);
  }
  ```
Login with cookie :  
  
![login cookie](https://github.com/JonPSmith/AuthPermissions.AspNetCore/assets/150670966/d76a0231-6a9a-4ad8-8a89-35c961e4be88)

The result :

![TenantList](https://github.com/JonPSmith/AuthPermissions.AspNetCore/assets/150670966/32505dfb-daaa-403d-9453-b91bf8e69baf)

Now it will be easier to use Authp in blazor. Thanks to new feature of NET 8.  

Actually I am still very curious how to use authp in blazor ssr which is almost the same as razor pages.

**Comment by meliora-solution** [comment]
I know now why EmailConfirmed must be set to true to make it works. It is because I set it in program.cs.

```
builder.Services.AddIdentityCore<IdentityUser>(options => options.SignIn.RequireConfirmedAccount = true)
    .AddEntityFrameworkStores<ApplicationDbContext>()
    .AddSignInManager()
    .AddDefaultTokenProviders();
```

**Comment by meliora-solution** [comment]
It is still too early if there is a possibility to use blazor NET8 for implementing Authp. But so far so good. I only tried one function to Get TenantList.  

First I create webApp in NET 8  and choose Automatic render mode = none.  

## Login ##

![Login](https://github.com/JonPSmith/AuthPermissions.AspNetCore/assets/150670966/b3495007-14e7-45a6-b0ec-71c33bfa85a5)

## Result ##
Yes I know I still have a problem with TenantRolesNames - it shows only the collection. I will solve it latter.

![TenantListBlazor](https://github.com/JonPSmith/AuthPermissions.AspNetCore/assets/150670966/aa03637e-64c3-4e71-8d9e-5eb7869eb27f)

##Permission##

As you see from the picture to apply permission use :

```
@attribute [HasPermission(Example3Permissions.TenantList)]

```
![Home](https://github.com/JonPSmith/AuthPermissions.AspNetCore/assets/150670966/621014fa-e63c-436d-b331-6089b1aa6da1)

**Comment by JonPSmith** [comment]
hi @meliora-solution,

Good progress. Looks like you found all the parts to make your multi-tenant app with a blazor frontend.

**Reply by meliora-solution** [reply]
Thanks to NET 8 features actually.  But I still have a problem to use Authp with customize identity.  I create ApplicationUser inherit from Identity User.   Is that even possible to use Authp with customize IdentityUser  ?

**Comment by meliora-solution** [comment]
I find a way how to add Additional Field in IdentityUser. I just want to add DateCreated to IdentityUser.  So I just add it to The Migration.

```
  DateCreated = table.Column<DateTime>(type: "DateTime", nullable: true),

```
